Source/WebKit/WebProcess/cocoa/TextTrackRepresentationCocoa.mm

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+
+#if ENABLE(VIDEO_PRESENTATION_MODE)
+#import "TextTrackRepresentationCocoa.h"
+
+#import "VideoFullscreenManagerProxyMessages.h"
+#import "WebPage.h"
+
+#import <WebCore/GraphicsContext.h>
+
+namespace WebKit {
+
+TextTrackRepresentationCocoa::TextTrackRepresentationCocoa(WebCore::TextTrackRepresentationClient& client)
+    : WebCore::TextTrackRepresentationCocoa(client)
+{
+}
+
+void TextTrackRepresentationCocoa::update()
+{
+    if (!m_page)
+        return;
+    auto representation = m_client.createTextTrackRepresentationImage();
+    if (!representation)
+        return;
+    auto image = representation->nativeImage();
+    if (!image)
+        return;
+    auto imageSize = image->size();
+    auto bitmap = ShareableBitmap::create(image->size(), { image->colorSpace() });
+    if (!bitmap)
+        return;
+    auto context = bitmap->createGraphicsContext();
+    if (!context)
+        return;
+    context->drawNativeImage(*image, imageSize, WebCore::FloatRect({ }, imageSize), WebCore::FloatRect({ }, imageSize), { WebCore::CompositeOperator::Copy });
+    auto handle = bitmap->createHandle();
+    if (!handle)
+        return;
+    m_page->send(Messages::VideoFullscreenManagerProxy::TextTrackRepresentationUpdate(*handle));
```

Comment:
It seems kind of weird to send messages directly to the VideoFullscreenManagerProxy from someplace that's not the VideoFullscreenManager. Instead of passing in `WebPage`, perhaps we could pass in `VideoFullscreenManager` (which WebPage owns) and talk directly to it.

Reply:
It still seems very weird to be sending messages directly to the `VideoFullscreenManagerProxy` via the page. You're also not sending a `contextId` in the message, which means there's no way to type the representation to any particular VideoFulllscreenInterface. 

I would say,`VideoFullscreenManager` needs to have the following methods:

```
void updateTextTrackRepresentationForVideoElement(WebCore::HTMLVideoElement&, const ShareableBitmapHandle&);
void setTextTrackRepresentationContentScaleForVideoElement(WebCore::HTMLVideoElement&, float scale);
void setTextTrackRepresentationIsHiddenForVideoElement(WebCore::HTMLVideoElement&, bool hidden);
```

Each of these methods would actually do the sending, after deriving a `contentId` from the passed in element.

Then, the equivalent methods in VideoFullscreenManagerProxy would pass the message contents on to VideoFullscreenInterfaceAVKit, which would actually insert the contents into the layer hierarchy.